subsys/emds/emds.c

```diff
@@ -6,55 +6,320 @@
 
 #include <emds/emds.h>
 
+#include <zephyr.h>
+#include <storage/flash_map.h>
+#include <drivers/flash.h>
+#include "emds_flash.h"
+
 #include <logging/log.h>
 LOG_MODULE_REGISTER(emds, CONFIG_EMDS_LOG_LEVEL);
 
+#define FLASH_TIME_PER_WORD_US (CONFIG_EMDS_FLASH_TIME_WRITE_ONE_WORD_US)
+#define FLASH_ENTRY_TIME_OVERHEAD_US (CONFIG_EMDS_FLASH_TIME_ENTRY_OVERHEAD_US)
+#define FLASH_BASE_TIME_OVERHEAD_US (CONFIG_EMDS_FLASH_TIME_BASE_OVERHEAD_US)
+
+K_SEM_DEFINE(emds_sem, 0, 1);
+static bool emds_ready;
+static bool emds_initialized;
+
+static sys_slist_t emds_dynamic_entries;
+static struct emds_fs emds_flash;
+
+static void emds_handler(void)
+{
+	k_sem_reset(&emds_sem);
+	k_sem_take(&emds_sem, K_FOREVER);
+
+	k_sched_lock();
+
+	LOG_DBG("Emergency Data Storeage released");
+
+	STRUCT_SECTION_FOREACH(emds_entry, ch) {
+		ssize_t len = emds_flash_write(&emds_flash,
+					ch->id, ch->data, ch->len);
+		if (len < 0) {
+			LOG_ERR("Write static entry: (%d) error (%d)",
+				ch->id, len);
+		} else if (len != ch->len) {
+			LOG_ERR("Write static entry: (%d) failed (%d:%d)",
+				ch->id, ch->len, len);
+		}
+	}
+
+	struct emds_dynamic_entry *ch;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		ssize_t len = emds_flash_write(&emds_flash,
+					ch->entry.id, ch->entry.data, ch->entry.len);
+		if (len < 0) {
+			LOG_ERR("Write dynamic entry: (%d) error (%d).",
+				ch->entry.id, len);
+		}
+		if (len != ch->entry.len) {
+			LOG_ERR("Write dynamic entry: (%d) failed (%d:%d).",
+				ch->entry.id, ch->entry.len, len);
+		}
+	}
+
+	k_sem_give(&emds_sem);
```

Comment:
Why do you give a semaphore counter twice for semaphore with counter limitation 1? 
Does it have the sense at all?

Reply:
The first take will hang because the take is waiting that the semaphore is going to be 1. So when the emds_store function does a give it will release the semaphore hanging in the thread. The semaphore will then be 0 again. So the emds_sem count is always zero until this is done. I tested this by setting up k_sem_count_get different places, and if we are going to use that as a check for when the storage is finished we need to do a give here, or it will always be zero. So it is this or using a bool value.

Reply:
Semaphore counter is initialized here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/sem.c#L53-L54 and initially counter is zero.
Then `k_sem_take` just lock spin but the counter is still zero. 
https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/sem.c#L124
Then `k_sem_give` will inlock thread without changing counter 
https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/sem.c#L97
Then  `k_sem_give` will add counter +1 since you provide the ability to not lock the next thread since you have some abstract source here: 
https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/sem.c#L99

I'm pretty sure you try to use semaphore and its counter not how it has been assumed to be used by design.

Reply:
please use `bool` instead